pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/BaseTaskExecutor.java

```diff
@@ -101,4 +106,29 @@ private void addTaskMeterMetrics(MinionMeter meter, long unitCount, String table
     _minionMetrics.addMeteredTableValue(tableName, meter, unitCount);
     _minionMetrics.addMeteredTableValue(tableName, taskType, meter, unitCount);
   }
+
+  protected void downloadSegmentToLocal(String tableNameWithType, String segmentName, String deepstoreURL,
+      String taskType, File tarredSegmentFile, String crypterName)
+      throws Exception {
+    LOGGER.info("Downloading segment from {} to {}", deepstoreURL, tarredSegmentFile.getAbsolutePath());
+    try {
+      // download from deepstore first
+      SegmentFetcherFactory.fetchAndDecryptSegmentToLocal(deepstoreURL, tarredSegmentFile, crypterName);
+    } catch (Exception e) {
+      LOGGER.error("Segment download failed from deepstore for {}, crypter:{}", deepstoreURL, crypterName, e);
+      TableConfig tableConfig = getTableConfig(tableNameWithType);
+      String peerDownloadScheme = tableConfig.getValidationConfig().getPeerSegmentDownloadScheme();
+      if (MinionTaskUtils.extractMinionAllowDownloadFromServer(tableConfig, taskType) && peerDownloadScheme != null) {
+        LOGGER.info("Trying to download from servers for segment {} post deepstore download failed", segmentName);
+        SegmentFetcherFactory.getSegmentFetcher(
+                getTableConfig(tableNameWithType).getValidationConfig().getPeerSegmentDownloadScheme())
+            .fetchSegmentToLocal(segmentName, () ->
+                PeerServerSegmentFinder.getPeerServerURIs(MINION_CONTEXT.getHelixManager(),
```

Comment:
one final nit: let's shuffle URIs here. Since the uriSupplier based method doesn't shuffle itself across retries. it lets the supplier control that behavior.

Reply:
since we are going to iterate through the entire list anyways does it make sense to shuffle?

Reply:
Yeah, because the supplier.get call is made for each retry. see `BaseSegmentFetcher#fetchSegmentToLocal`. If you shuffle each time, if there's a bad server you will likely skip it in one of retries.

Reply:
yeah makes sense! Updated.